C#: AST printing

tamasvajk opened the pull request:
Fixes https://github.com/github/codeql/issues/3957

hvitved commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -110,9 +110,16 @@ private predicate isCompilerGeneratedAttributable(Attributable attributable) {
+ * Retrieves the canonical QL class(es) for entity `el`
  */
-private string getQlClass(Element el) { result = "[" + concat(el.getAQlClass(), ",") + "] " }
+private string getQlClass(Element el) {
+  if count(el.getAPrimaryQlClass()) = 1 and not el.getAPrimaryQlClass() = "???"
+  then result = "[" + el.getAPrimaryQlClass() + "] "
+  else
+    if count(el.getAPrimaryQlClass()) > 1
+    then result = "[" + concat(el.getAPrimaryQlClass(), ",") + "] "
+    else result = "ERR [" + concat(el.getAQlClass(), ",") + "] "
+}
```
This will force computation of `getAQlClass()`, which is often very expensive. I therefore think we should do the same as C++, i.e.

```
/**
 * Retrieves the canonical QL class(es) for entity `el`
 */
private string qlClass(ElementBase el) {
  result = "[" + concat(el.getAPrimaryQlClass(), ",") + "] "
  // Alternative implementation -- do not delete. It is useful for QL class discovery.
  //result = "["+ concat(el.getAQlClass(), ",") + "] "
}
```

hvitved commented on csharp/ql/src/semmle/code/csharp/Property.qll:
```diff
@@ -370,6 +374,10 @@ class Accessor extends Callable, Modifiable, Attributable, @callable_accessor {
 
   override string toString() { result = getName() }
+
+  override string getAPrimaryQlClass() {
+    result = "Accessor" and not this instanceof Getter and not this instanceof Setter
+  }
```
I don't think this is needed as the other accessors are `AddEventAccessor` and `RemoveEventAccessor` which define their own `getAPrimaryQlClass()`.

hvitved commented on csharp/ql/src/semmle/code/csharp/Stmt.qll:
```diff
@@ -1171,6 +1227,12 @@ class LocalVariableDeclStmt extends Stmt, @decl_stmt {
   override string toString() { result = "... ...;" }
+
+  override string getAPrimaryQlClass() {
+    result = "LocalConstantDeclStmt" and
```
This should be `LocalVariableDeclStmt`, right?

calumgrant commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -0,0 +1,606 @@
+ * Printed AST nodes are mostly `Element`s of the underlying AST.
+ * There are extra AST nodes generated for parameters of `Parameterizable`s,
+ * attributes of `Attributable`s, and type parameters of `UnboundGeneric`
+ * types. These extra node are used as containers to organize the tree a
```
typo `nodes`

calumgrant commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -0,0 +1,606 @@
+
+  AttributesNode() {
+    this = TAttributesNode(attributable) and
+    count(attributable.getAnAttribute()) > 0 and
```
`exists(attributable.getAnAttribute())` might be clearer

hvitved commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -0,0 +1,606 @@
+  isInsideUnneededType(attributable.(Field).getDeclaringType()) or
+  isInsideUnneededParameterizable(attributable.(Parameter).getDeclaringElement()) or
+  isInsideUnneededCallable(attributable.(Callable)) or
+  isInsideUnneededType(attributable.(DeclarationWithAccessors).getDeclaringType()) or
+  isInsideUnneededType(attributable.(ValueOrRefType))
+}
+
+private predicate isInsideUnneededUnboundGeneric(UnboundGeneric unboundGeneric) {
+  isInsideUnneededAttributable(unboundGeneric) or
+  isInsideUnneededParameterizable(unboundGeneric)
+}
+
+private predicate isInsideUnneededCallable(Callable callable) {
+  callable instanceof ConstructedGeneric or
+  isInsideUnneededType(callable.getDeclaringType())
+}
+
+private predicate isInsideUnneededType(Type type) {
+  type instanceof ConstructedType or
+  type.getDeclaringType*() instanceof ConstructedType or
+  type instanceof AnonymousClass or
+  type.getDeclaringType*() instanceof AnonymousClass
+}
+
+private predicate isInsideUnneededParameterizable(Parameterizable parameterizable) {
+  isInsideUnneededCallable(parameterizable) or
+  isInsideUnneededType(parameterizable.(Indexer).getDeclaringType()) or
+  isInsideUnneededType(parameterizable.(DelegateType))
+}
+
+private predicate isCompilerGeneratedParameterizable(Parameterizable parameterizable) {
+  parameterizable.isCompilerGenerated() or
+  parameterizable.getDeclaringType*().isCompilerGenerated()
+}
+
+private predicate isCompilerGeneratedAttributable(Attributable attributable) {
+  // Default parameter values on delegates are in the tree multiple times
+  // due to the compiler generated `Invoke`.
+  isCompilerGeneratedParameterizable(attributable.(Parameter).getDeclaringElement())
+}
```
I wonder if all of this could be simplified to:
```ql
predicate isNotNeeded(Element e) {
  e.(Expr).isImplicit()
  or
  e.(Declaration).isCompilerGenerated()
  or
  e instanceof ConstructedGeneric
  or
  e instanceof AnonymousClass
  or
  isNotNeeded(e.getParent())
  or
  isNotNeeded(e.(Attribute).getTarget())
}
```

hvitved replied:
Performance can also be improved by restricting to elements `e` for which `shouldPrint(e)` holds.

tamasvajk replied:
It would be great if we could convert those predicates to a single more elegant, more readable one. I applied this change, but it results in way too many changes in the expected files. I'll investigate the changes in smaller batches.

One initial finding:
Adding the `.isImplicit()` removes too many things from the trees:
* implicit `this` access, which is okay,
* implicit casts, which is definitely not great, because `10` would also be removed from `byte b = 10;`
* `OperatorCall` to implicit conversion

and maybe others too. So I'll omit this change, and instead I'll probably only handle implicit `this` access. What do you think?

hvitved replied:
Thanks a lot for investigating this; I agree that `isImplicit()` cannot be used then.

tamasvajk replied:
I pushed a fix for this. The following seems to cover the same cases as the initial PR:
```
private predicate isFilteredCompilerGenerated(Declaration d) {
  d.isCompilerGenerated() and
  not d instanceof Accessor
}

private predicate isNotNeeded(Element e) {
  isFilteredCompilerGenerated(e)
  or
  e instanceof ConstructedGeneric
  or
  e instanceof AnonymousClass
  or
  isNotNeeded(e.(Declaration).getDeclaringType())
  or
  isNotNeeded(e.(Parameter).getDeclaringElement())
  or
  isNotNeeded(e.(Attribute).getTarget())
}
```

hvitved replied:
Excellent @tamasvajk !

hvitved commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -0,0 +1,606 @@
+    )
+}
+
+private predicate locationSortKeys(Element ast, string file, int line, int column) {
+  if exists(getRepresentativeLocation(ast))
+  then
+    exists(Location loc |
+      loc = getRepresentativeLocation(ast) and
+      file = loc.getFile().toString() and
+      line = loc.getStartLine() and
+      column = loc.getStartColumn()
+    )
+  else (
+    file = "" and
+    line = 0 and
+    column = 0
+  )
+}
```
This can be simplified to
```ql
  exists(Location loc |
    loc = getRepresentativeLocation(ast) and
    file = loc.getFile().toString() and
    line = loc.getStartLine() and
    column = loc.getStartColumn()
  )
  or
  not exists(getRepresentativeLocation(ast)) and
  file = "" and
  line = 0 and
  column = 0
```

hvitved commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -0,0 +1,606 @@
+ * bit better.
+ */
+private newtype TPrintAstNode =
+  TAstNode(Element ast) { shouldPrint(ast) } or
```
Rename to `TElementNode`?

hvitved commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -0,0 +1,606 @@
+  abstract PrintAstNode getChild(int childIndex);
+
+  /**
+   * Gets the children of this node.
```
Gets a child

hvitved commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -0,0 +1,606 @@
+  override Location getLocation() { result = getRepresentativeLocation(ast) }
+
+  /**
+   * Gets the AST represented by this node.
```
Gets the element
And maybe rename to `getElement()`

hvitved commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -0,0 +1,606 @@
+        ao.getExpandedAssignment().getRValue() = controlFlowElement or
+        ao.getExpandedAssignment().getRValue().(BinaryOperation).getLeftOperand() =
+          controlFlowElement.getParent*() or
+        ao.getExpandedAssignment().getRValue().(DynamicOperatorCall).getChild(0) =
+          controlFlowElement.getParent*() or
```
This case should be covered by the case below.

hvitved commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -0,0 +1,606 @@
+    not isInsideUnneededType(declaration.getDeclaringType())
+  }
+
+  override PrintAstNode getChild(int childIndex) {
```
Same comment as for `TypeNode::getChild()`

hvitved commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -0,0 +1,606 @@
+    or
+    childIndex = 1 and
+    field.hasInitializer() and
+    (
+      not field.getDeclaringType() instanceof Enum and
+      result.(AstNode).getAst() = field.getInitializer().getParent()
+      or
+      field.getDeclaringType() instanceof Enum and
+      result.(AstNode).getAst() = field.getInitializer()
+    )
```
Here you may as well use `if-then-else`, i.e.
```
      if field.getDeclaringType() instanceof Enum
      then result.(AstNode).getAst() = field.getInitializer()
      else result.(AstNode).getAst() = field.getInitializer().getParent()
```

hvitved commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -0,0 +1,606 @@
+    result.(AttributesNode).getAttributable() = param
+    or
+    childIndex = 1 and
+    param.hasDefaultValue() and
```
This line is not needed

hvitved commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -0,0 +1,606 @@
+
+  ParametersNode() {
+    this = TParametersNode(parameterizable) and
+    parameterizable.getNumberOfParameters() > 0 and
+    not isInsideUnneededParameterizable(parameterizable) and
+    (
+      not parameterizable.isCompilerGenerated() or
+      parameterizable instanceof Accessor
+    )
```
You may as well move these constraints into the body of the `TParametersNode` `newtype`-branch.

tamasvajk replied:
Thanks. Is there any benefit of doing this? Other than having all constraints in one place? 

I guess it is possible to have multiple derived classes of a `newtype`-branch, and then I could have different restrictions on them. Is this a pattern that we use or we typically have one derived type of a `newtype`-branch, and then restrict that further by subtyping the class?

hvitved replied:
The benefit is that the implicit predicate generated by the QL compiler for the `TParametersNode` injector becomes smaller. Of course this only works, as you mention, if all classes that derive from the injector apply the same constraints.

tamasvajk replied:
I understand most part of what's written here. :-) What is an injector?

hvitved replied:
Sorry, not sure if it is standard terminology, but it is branch in a `newtype`.

hvitved commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -0,0 +1,606 @@
+  }
+
+  /**
+   * Returns the underlying `Parameterizable`
```
Returns -> Gets (see also https://github.com/github/codeql/blob/master/docs/qldoc-style-guide.md).

hvitved commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -0,0 +1,606 @@
+  Attributable attributable;
+
+  AttributesNode() {
+    this = TAttributesNode(attributable) and
+    exists(attributable.getAnAttribute()) and
+    not isCompilerGeneratedAttributable(attributable) and
+    not isInsideUnneededAttributable(attributable)
```
Again, move these up to `TAttributesNode`.

hvitved commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -0,0 +1,606 @@
+  }
+
+  /**
+   * Returns the underlying `Attributable`
```
Gets.

hvitved commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -0,0 +1,606 @@
+  UnboundGeneric unboundGeneric;
+
+  TypeParametersNode() {
+    this = TTypeParametersNode(unboundGeneric) and
+    unboundGeneric.getNumberOfTypeParameters() > 0 and
+    not isInsideUnneededUnboundGeneric(unboundGeneric)
```
Move up to `TTypeParametersNode`.

hvitved commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -0,0 +1,606 @@
+  }
+
+  /**
+   * Returns the underlying `UnboundGeneric`
```
Gets

tamasvajk commented on csharp/ql/src/semmle/code/csharp/PrintAst.ql:
```diff
@@ -0,0 +1,9 @@
+/**
+ * @name Print AST
+ * @description Outputs a representation of the Abstract Syntax Tree.
+ * @id csharp/print-ast
+ * @kind graph
+ */
+
+import csharp
+import PrintAst
```
All the tests are referencing this file. Do you think the tests should reference the top level VS Code specific `PrintAst.ql`?

hvitved replied:
Ah yes, let's keep it then.

hvitved replied:
Perhaps we should add a tweakable config then, as in [C++](https://github.com/github/codeql/blob/master/cpp/ql/src/semmle/code/cpp/PrintAST.ql#L11-L21)?

tamasvajk commented on csharp/ql/src/semmle/code/csharp/exprs/Dynamic.qll:
```diff
@@ -309,6 +317,8 @@ class DynamicAccessorCall extends DynamicAccess {
     )
   }
+
+  override string getAPrimaryQlClass() { result = "DynamicAccessorCall" }
```
I'm removing this. At the same time, I don't know anything about query stages. Is there some documentation or paper that I could learn from?

tamasvajk commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -0,0 +1,606 @@
+   */
+  string toString() { result = "PrintAstConfiguration" }
+
+  /**
+   * Controls whether the `File` should be considered for AST printing.
+   */
+  predicate selectedFile(File file) { any() }
+
+  /**
+   * Controls whether the `Element` should be considered for AST printing.
+   * By default it checks which file the `elem` belongs to.
+   */
+  predicate shouldPrint(Element elem) {
+    elem.fromSource() and
+    this.selectedFile(getRepresentativeLocation(elem).getFile())
+  }
```
What would be the implementation of the `private predicates` in this case? 
```
private predicate shouldPrint(Element e) {
  exists(PrintAstConfiguration config | config.shouldPrint(e, e.getFile()))
}

private predicate selectedFile(File file) {
  exists(PrintAstConfiguration config | config.shouldPrint(_, file))
}
```

hvitved replied:
Perhaps those two should be replaced by a single `shouldPrint(Element e, File f)` predicate as well.

tamasvajk commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -0,0 +1,606 @@
+    not isInsideUnneededType(type)
+  }
+
+  override PrintAstNode getChild(int childIndex) {
+    childIndex = 0 and
+    result.(AttributesNode).getAttributable() = type
+    or
+    childIndex = attributableOffset(type) and
+    result.(TypeParametersNode).getUnboundGeneric() = type
+    or
+    childIndex = attributableOffset(type) + unboundGenericOffset(type) and
+    result.(ParametersNode).getParameterizable() = type
+    or
+    result.(AstNode).getAst() =
+      rank[childIndex - attributableOffset(type) - unboundGenericOffset(type) -
+          parameterizableTypeOffset(type)](Member m, string file, int line, int column |
+        m = type.getAMember() and
+        locationSortKeys(m, file, line, column)
+      |
+        m order by file, line, column
+      )
+  }
```
This is awesome, thanks for catching this. I didn't like these `*Offset` predicates.

hvitved commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -31,46 +31,27 @@ private predicate shouldPrint(Element e, File f) {
-  isInsideUnneededType(attributable.(DeclarationWithAccessors).getDeclaringType()) or
-  isInsideUnneededType(attributable.(ValueOrRefType))
+private predicate isImplicitExpression(ControlFlowElement element) {
+  element instanceof Expr and element.(Expr).isImplicit() and not exists(element.getAChild())
```
`element instanceof Expr` is not needed as it is implied from the infix cast in `element.(Expr).isImplicit()`.

hvitved commented:
I ran a quick performance experiment by evaluating `PrintAst.ql` on `mono`, and it revealed a problematic join-order:
```
[2020-08-17 09:21:47] (868s) Tuple counts for PrintAst::getRepresentativeLocation#ff#shared:
                      563439     ~0%     {5} r1 = SCAN Location::Location::hasLocationInfo_dispred#ffffff AS I OUTPUT I.<0>, I.<2>, I.<3>, I.<4>, I.<5>
                      563439     ~0%     {6} r2 = JOIN r1 WITH Location::Location::getFile_dispred#ff AS R ON FIRST 1 OUTPUT R.<1>, r1.<0>, r1.<1>, r1.<2>, r1.<3>, r1.<4>
                      1016722416 ~0%     {6} r3 = JOIN r2 WITH PrintAst::shouldPrint#ff_10#join_rhs AS R ON FIRST 1 OUTPUT R.<1>, r2.<1>, r2.<2>, r2.<3>, r2.<4>, r2.<5>
                                         return r3
```
What happens above is join of all `Location`s with with `Element`s that agree on the `File`, which is incomprehensible to compute. One possible fix is https://github.com/hvitved/ql/commit/04f38c48d148d0f4e99265b0f26759469aaa007c, which you may cherry-pick, but another approach is to change the signature of `PrintAstConfiguration::shouldPrint(Element e, File f)` to `PrintAstConfiguration::shouldPrint(Element e, Location l)`. I think I prefer the latter solution.

hvitved commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -0,0 +1,530 @@
+ *
+ * By default, this will print the AST for all elements in the database. To change this behavior,
+ * extend `PrintAstConfiguration` and override `shouldPrint` to hold for only the elements
+ * you wish to view the AST for, or override `selectedFile` to hold for only the files you
```
The part about `selectedFile` should be removed now.

tamasvajk commented:
@hvitved I pushed the fixes for your findings. I've split the "base type" printing into two commits, the first one is putting the element at the bottom of the declarations, this way it's easier to check the `.expected` files. The second commit moves the tree nodes to the correct locations, which means changes to all declaration members.

hvitved commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -112,6 +132,18 @@ private newtype TPrintAstNode =
+    not isNotNeeded(type)
+  } or
+  TBaseTypeNode(ValueOrRefType derived, ValueOrRefType base) {
+    shouldPrint(derived, _) and
+    hasInterestingBaseTypes(derived) and
+    not isNotNeeded(derived) and
+    derived.getABaseType() = base and
+    isInterestingBaseType(base)
```
Can be simplified to
```ql
    shouldPrint(derived, _) and
    base = getAnInterestingBaseType(derived) and
    not isNotNeeded(derived)
```

hvitved commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -504,6 +539,66 @@ final class TypeParametersNode extends PrintAstNode, TTypeParametersNode {
+
+  BaseTypesNode() { this = TBaseTypesNode(valueOrRefType) }
+
+  override string toString() { result = "(BaseTypes)" }
```
Perhaps `"(Base types)"`

hvitved commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -504,6 +539,66 @@ final class TypeParametersNode extends PrintAstNode, TTypeParametersNode {
+  override Location getLocation() { none() }
+
+  override BaseTypeNode getChild(int childIndex) {
+    childIndex = 0 and
+    result.getBaseType() = valueOrRefType.getBaseClass() and
+    isInterestingBaseType(valueOrRefType.getBaseClass()) and
+    result.getDerivedType() = valueOrRefType
```
Since the relevant constraints are already applied to `TBaseTypeNode`, you can remove the `isInterestingBaseType(valueOrRefType.getBaseClass())` conjunct.

hvitved commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -504,6 +539,66 @@ final class TypeParametersNode extends PrintAstNode, TTypeParametersNode {
+    result.getDerivedType() = valueOrRefType
+    or
+    result.getBaseType() =
+      rank[childIndex - 1](ValueOrRefType base, string name |
```
Can just as well be `childIndex`, as ranking always starts from 1.

hvitved commented on csharp/ql/src/semmle/code/csharp/PrintAst.qll:
```diff
@@ -0,0 +1,623 @@
+
+  BaseTypeNode() { this = TBaseTypeNode(derived, base) }
+
+  override string toString() { result = base.toString() }
```
Should perhaps be `result = getQlClass(base) + base`.

tamasvajk commented:
note to self: do not merge this without regenerating the expected files. https://github.com/github/codeql/pull/4055 and https://github.com/github/codeql/pull/4094 are changing them.

tamasvajk commented:
@hvitved I rebased, adjusted the expected files in a new commit, and force pushed the branch. Let's wait for the checks to finish, and then we can squash-merge this.

aeisenberg commented:
🎉  Thanks for getting this done, @tamasvajk!
